packages/kinvey-js-sdk/src/files.js

```diff
@@ -0,0 +1,224 @@
+  });
+  const response = await execute(request);
+  return response.data;
+}
+
+function checkUploadStatus(url, headers, metadata, timeout) {
+  const requestHeaders = new Headers(headers);
+  requestHeaders.set('Content-Type', metadata.mimeType);
+  requestHeaders.set('Content-Range', `bytes */${metadata.size}`);
+  const request = new Request({
+    method: RequestMethod.PUT,
+    headers: requestHeaders,
+    url,
+    timeout
+  });
+  return execute(request, true);
+}
+
+function getStartIndex(rangeHeader, max) {
+  const start = rangeHeader ? parseInt(rangeHeader.split('-')[1], 10) + 1 : 0;
+  return start >= max ? max - 1 : start;
+}
+
+function randomInt(min, max) {
+  return Math.floor(Math.random() * (max - min)) + min;
+}
+
+async function uploadFile(url, file, metadata, options) {
+  const { count = 0, maxBackoff = MAX_BACKOFF } = options;
+  let { start = 0 } = options;
+
+  const requestHeaders = new Headers(options.headers);
+  requestHeaders.set('Content-Type', metadata.mimeType);
+  requestHeaders.set('Content-Range', `bytes ${options.start}-${metadata.size - 1}/${metadata.size}`);
+  const request = new Request({
+    method: RequestMethod.PUT,
+    headers: requestHeaders,
+    url,
+    body: file.slice(options.start, metadata.size)
+  });
+  const response = await execute(request, true);
+
+  if (!response.isSuccess()) {
+    // An error occurred. We should retry uploading the file after some time has passed
+    if (response.statusCode >= 500 && response.statusCode < 600) {
+      const backoff = (2 ** options.count) + randomInt(1, 1001); // Calculate the exponential backoff
+
+      if (backoff < options.maxBackoff) {
+        await new Promise((resolve) => {
+          setTimeout(resolve, backoff);
+        });
+        return uploadFile(url, file, metadata, { count: count + 1, start, maxBackoff });
+      }
+    }
+
+    throw response.error;
+  }
+
+  // The upload isn't complete and we must upload the rest of the file
+  if (response.statusCode === 308) {
+    start = getStartIndex(response.headers.get('Range'), metadata.size);
+    return uploadFile(url, file, metadata, { count: 0, start, maxBackoff });
+  }
+
+  return response.data;
+}
+
+export async function upload(file = {}, metadata = {}, options = {}) {
+  const fileMetadata = transformMetadata(file, metadata);
+  const kinveyFile = await saveFileMetadata(fileMetadata, options);
+  const uploadStatusResponse = await checkUploadStatus(kinveyFile._uploadURL, kinveyFile._requiredHeaders, fileMetadata, options.timeout);
+
+  if (!uploadStatusResponse.isSuccess()) {
+    throw uploadStatusResponse.error;
+  }
+
+  if (uploadStatusResponse.statusCode !== 200 && uploadStatusResponse.statusCode !== 201) {
+    if (uploadStatusResponse.statusCode !== 308) {
+      // TODO: Here we should handle redirects according to location header, but this generally shouldn't happen
+      throw new Error('Unexpected response for upload file status check request.');
```

Comment:
Does it make sense here to also return the `statusCode` that caused this exception, since this throw is for any error received?

Reply:
I will add log statements that will contain more details. We have this today and I need to do it here.